lib/src/fixed_datetime_formatter.dart

```diff
@@ -0,0 +1,184 @@
+  FixedDateTimeFormatter(this.pattern) {
+    String? current;
+    var start = 0;
+    for (var i = 0; i < pattern.length; i++) {
+      var char = pattern[i];
+      if (current != char) {
+        _saveFormatBlock(current, start, i);
+        if (_validFormatCharacters.contains(char)) {
+          var hasSeenBefore = _occurences.containsKey(char);
+          if (hasSeenBefore) {
+            throw FormatException(
+                "Pattern contains more than one '$char' block.\n"
+                "Previous occurrence at position ${_occurences[char]!.from}",
+                pattern,
+                i);
+          } else {
+            start = i;
+            current = char;
+          }
+        } else {
+          current = null;
+        }
+      }
+    }
+    _saveFormatBlock(current, start, pattern.length);
+  }
+
+  void _saveFormatBlock(String? current, int start, int i) {
+    if (current != null) _occurences[current] = _Range(start, i);
+  }
+
+  /// Convert [DateTime] to a [String] exactly as specified by the [pattern].
+  String encode(DateTime datetime) {
+    var buffer = StringBuffer();
+    var previousEnd = 0;
+    _occurences.forEach((key, value) {
+      if (previousEnd < value.from) {
+        buffer.write(pattern.substring(previousEnd, value.from));
+      }
+      var number = _extractNumFromDateTime(key, datetime).toString();
+      var length = value.length;
+      if (number.length > length) {
+        number = number.substring(number.length - length);
+      } else if (number.length < length) {
+        number = number.padLeft(length, '0');
+      }
+      buffer.write(number);
+      previousEnd = value.to;
+    });
+    if (previousEnd < pattern.length) {
+      buffer.write(pattern.substring(previousEnd, pattern.length));
+    }
+    return buffer.toString();
+  }
+
+  int _extractNumFromDateTime(String? key, DateTime dt) {
+    switch (key) {
+      case 'Y':
+        return dt.year;
+      case 'C':
+        return (dt.year / 100).floor();
+      case 'E':
+        return (dt.year / 10).floor();
+      case 'M':
+        return dt.month;
+      case 'D':
+        return dt.day;
+      case 'h':
+        return dt.hour;
+      case 'm':
+        return dt.minute;
+      case 's':
+        return dt.second;
+    }
+    throw AssertionError("Unreachable, the key is checked in the constructor");
+  }
+
+  /// Parse a string [formattedDateTime] to a local [DateTime] as specified in the
+  /// [pattern], substituting missing values with a default. Throws an exception
+  /// on failure to parse.
+  DateTime decode(String formattedDateTime, [bool isUtc = false]) {
+    return _decode(formattedDateTime, isUtc, int.parse);
+  }
+
+  /// Same as [decode], but will not throw on parsing erros, instead using
+  /// the default value as if the format char was not present in the [pattern].
+  DateTime tryDecode(String formattedDateTime, [bool isUtc = false]) {
```

Comment:
Make `isUtc` a named parameter. It's recommended for boolean parameters in public APIs.